applications/RomApplication/custom_utilities/rom_residuals_utility.h

```diff
@@ -0,0 +1,185 @@
+#include "solving_strategies/schemes/scheme.h"
+#include "spaces/ublas_space.h"
+
+/* Application includes */
+#include "rom_application_variables.h"
+
+namespace Kratos
+{
+    typedef UblasSpace<double, CompressedMatrix, boost::numeric::ublas::vector<double>> SparseSpaceType;
+    typedef UblasSpace<double, Matrix, Vector> LocalSpaceType;
+    typedef Scheme<SparseSpaceType, LocalSpaceType> BaseSchemeType;
+
+    // This utility returns the converged residuals projected onto the ROM basis Phi.
+    class RomResidualsUtility
+    {
+        public:
+
+        KRATOS_CLASS_POINTER_DEFINITION(RomResidualsUtility);
+
+        RomResidualsUtility(
+        ModelPart& rModelPart,
+        Parameters ThisParameters,
+        BaseSchemeType::Pointer pScheme
+        ): mpModelPart(rModelPart), mpScheme(pScheme){
+        // Validate default parameters
+        Parameters default_parameters = Parameters(R"(
+        {
+            "nodal_unknowns" : [],
+            "number_of_rom_dofs" : 10
+        })" );
+
+        ThisParameters.ValidateAndAssignDefaults(default_parameters);
+
+        mNodalVariablesNames = ThisParameters["nodal_unknowns"].GetStringArray();
+
+        mNodalDofs = mNodalVariablesNames.size();
+        mRomDofs = ThisParameters["number_of_rom_dofs"].GetInt();
+
+        // Setting up mapping: VARIABLE_KEY --> CORRECT_ROW_IN_BASIS
+        for(int k=0; k<mNodalDofs; k++){
+            if(KratosComponents<Variable<double>>::Has(mNodalVariablesNames[k]))
+            {
+                const auto& var = KratosComponents<Variable<double>>::Get(mNodalVariablesNames[k]);
+                MapPhi[var.Key()] = k;
+            }
+            else if(KratosComponents<ModelPart::VariableComponentType>::Has(mNodalVariablesNames[k]))
+            {
+                const auto& var = KratosComponents<ModelPart::VariableComponentType>::Get(mNodalVariablesNames[k]);
+                MapPhi[var.Key()] = k;
+            }
+            else
+                KRATOS_ERROR << "variable \""<< mNodalVariablesNames[k] << "\" not valid" << std::endl;
+
+        }
+    }
+
+        ~RomResidualsUtility()= default;
+
+
+        void GetPhiElemental(
+            Matrix &PhiElemental,
+            const Element::DofsVectorType &dofs,
+            const Element::GeometryType &geom)
+        {
+            const auto *current_rom_nodal_basis = &(geom[0].GetValue(ROM_BASIS));
+            int counter = 0;
+            for(unsigned int k = 0; k < dofs.size(); ++k){
+                auto variable_key = dofs[k]->GetVariable().Key();
+                if(k==0)
+                    current_rom_nodal_basis = &(geom[counter].GetValue(ROM_BASIS));
+                else if(dofs[k]->Id() != dofs[k-1]->Id()){
+                    counter++;
+                    current_rom_nodal_basis = &(geom[counter].GetValue(ROM_BASIS));
+                }
+                if (dofs[k]->IsFixed())
+                    row(PhiElemental, k) = ZeroVector(PhiElemental.size2());
```

Comment:
noalias?

Reply:
In #6534 I will be adding the noalias. Also in that PR, since the structure is very similar, I am adding the paralelism to this residual utility